### PR TITLE
[SPARK-26222][SQL] Track file listing time

### PR DESCRIPTION
**Pull request by xuanyuanking**
## What changes were proposed in this pull request?

Tracking file listing time and add them into scan node's metrics, also add the start and end timestamp metrics for both file listing phase and partition pruning phase.

## How was this patch tested?

Add test in QueryPlanningTrackerEndToEndSuite and SQLMetricsSuite.
Manually test as below:

|             Scenario            | UI |
|:-------------------------------:|----|
| With file listing               |  image |
| Without file listing(cache hit) |  image |
| With partition pruning          | image |

### CODE REVIEW

**Comment by xuanyuanking** [comment]
retest this please.

**Comment by xuanyuanking** [comment]
retest this please.

**Comment by xuanyuanking** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala [hunk 1]
```diff
@@ -161,3 +166,8 @@ class QueryPlanningTracker {
   }
 
 }
+
+object NoopTracker extends QueryPlanningTracker {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala]: why is this needed?

Comment (rxin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala]: ic - you just don't want to deal with the Option ?

Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala]: Yea, you're right.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileIndex.scala [hunk 2]
```diff
@@ -72,14 +72,4 @@ trait FileIndex {
...
-
-  /**
-   * Returns an optional metadata operation time, in nanoseconds, for listing files.
-   *
-   * We do file listing in query optimization (in order to get the proper statistics) and we want
-   * to account for file listing time in physical execution (as metrics). To do that, we save the
-   * file listing time in some implementations and physical execution calls it in this method
-   * to update the metrics.
-   */
-  def metadataOpsTimeNs: Option[Long] = None
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileIndex.scala]: This is part of https://github.com/apache/spark/pull/23298/commits/73bf55389466a553606d1878848c0bb018b36dc2, here I revert the whole changes in [SPARK-20151](https://github.com/apache/spark/pull/17476/files). Maybe use the QueryPlanningTracker to measure the pruning phase will have a better code base in `FileSourceScanExec`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/QueryPlanningMetricsSupport.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+   * Get updated file listing relative metrics from QueryPlanningTracker phases.
+   */
+  def getUpdatedFileListingMetrics(metrics: Map[String, SQLMetric]): Seq[SQLMetric] = {
+    val updatedMetrics = new ArrayBuffer[SQLMetric]()
+
+    // Update all metrics relative with file listing phase.
+    def phaseMetricsUpdate(phase: String): Unit = {
+      val phaseSummary = QueryPlanningTracker.get.phases.get(phase)
+      if (phaseSummary.isDefined) {
+        metrics(FILE_LISTING_TIME).add(phaseSummary.get.durationMs)
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/QueryPlanningMetricsSupport.scala]: Here we also have another choice, which is to separate file listing time and partition pruning time, I just followed the original method of adding them together as file listing time.

**Comment by xuanyuanking** [comment]
It's a wrong way to add metrics like this. It can't handle the scenario of union and join.
